mlir/lib/Dialect/Rock/Transforms/ThreadwiseGemmLowering.cpp

```diff
@@ -89,6 +89,168 @@ struct ThreadwiseReadIntoRewritePattern
+
+    Location loc = op.getLoc();
+
+    Value bufferA = adaptor.getMatrixA();
+    Value bufferB = adaptor.getMatrixB();
+    Value bufferC = adaptor.getMatrixC();
+
+    auto dataTypeA = adaptor.getMatrixA().getType().cast<MemRefType>().getElementType();
+    auto dataTypeB = adaptor.getMatrixB().getType().cast<MemRefType>().getElementType();
+
+    auto bufferAShape = op.getMatrixA().getType().getShape();
+    auto bufferCShape = op.getMatrixC().getType().getShape();
+
+    if (dataTypeA.isa<VectorType>()) {
+        dataTypeA = dataTypeA.cast<VectorType>().getElementType();
+    }
+    if (dataTypeB.isa<VectorType>()) {
+        dataTypeB = dataTypeB.cast<VectorType>().getElementType();
+    }
+
+    auto emitter = rock::accel::AccelEmitter::select(
+          op.getFeatures(), dataTypeA, dataTypeB, op.getArch(), op.getParams());
+
+    if (!emitter)
+        return emitError(loc)
+              << "Failed to select any accelerator instruction.\n";
+
+    rock::accel::AccelEmitterParams params = emitter->getParams();
+    Type argTypeA = params.argTypeA;
+    Type argTypeB = params.argTypeB;
+
+    // llvm::outs() << "\n**** ARGTYPE: " << argTypeA << "\n";
+
+    TransformMapAttr normalizedViewA, normalizedViewB, normalizedViewC;
+
+    // Sizes of the [i,j,k] axis
+    int64_t i = *(bufferCShape.end() - 2);
+    int64_t j = bufferCShape.back();
+    int64_t k = bufferAShape.back();
+
+    if(isMfma || isWmma){
+      normalizedViewA =
+          normalizeView(b, loc, {"i", "j", "k"}, {i, j, k}, true, {"j"});
+      normalizedViewB =
+          normalizeView(b, loc, {"i", "j", "k"}, {i, j, k}, true, {"i"});
+      normalizedViewC = normalizeView(
+          b, loc, {"i", "j", "k"}, {i, j, k}, false, {"k"});
+    } else {
+      int64_t kPack = bufferAShape[2];
+
+      SmallVector<int64_t, 3> dimensions = {k, i, j, kPack};
+      SmallVector<int64_t, 3> strides = {1, 1, 1, 1};
+       
+      normalizedViewA = normalizeView(b, loc, {"k", "i", "j", "kpack"}, dimensions, true, {"j"});
+      normalizedViewB = normalizeView(b, loc, {"k", "i", "j", "kpack"}, dimensions, true, {"i"});
+      normalizedViewC = normalizeView(b, loc, {"k", "i", "j", "kpack"}, dimensions, true, {"k", "kpack"});
+    }
+
+    auto [rawBufferA, bufferViewA, sourceANeeds64BitIdx] =
+          untransform(b, bufferA, normalizedViewA);
+    auto [rawBufferB, bufferViewB, sourceBNeeds64BitIdx] =
+          untransform(b, bufferB, normalizedViewB);
+    auto [rawBufferC, bufferViewC, dstNeeds64BitIdx] =
+          untransform(b, bufferC, normalizedViewC);
+    assert(!sourceANeeds64BitIdx && "Registers shouldn't need 64-bit indexing");
+    assert(!sourceBNeeds64BitIdx && "Registers shouldn't need 64-bit indexing");
+    assert(!dstNeeds64BitIdx && "Registers shouldn't need 64-bit indexing");
+
+    // Loop properties
+    auto computeStart = llvm::to_vector(op.getComputeIndices());
+
+    if(isMfma || isWmma){
```

Comment:
Why? Don't WMMA and MFMA go down `threadwise_gemm`?

Reply:
This is temporary for parts that are not unified yet. The goal is to have it removed at the end of the unification paths.